lib/src/useTernaryDarkMode/useTernaryDarkMode.ts

```diff
@@ -1,10 +1,87 @@
-import {} from 'react'
+/* usage:
+ *   const { isPaletteDarkMode, ternaryModeCode, togglePaletteDarkMode } = usePaletteDarkMode();
+ */
 
-function useTernaryDarkMode(): [number, () => number] {
-  const value = 2
-  const method = () => 2
+import { useEffect, useState } from 'react'
 
-  return [value, method]
+// See: https://usehooks-ts.com/react-hook/use-local-storage
+import { useLocalStorage } from '../useLocalStorage'
+// See: https://usehooks-ts.com/react-hook/use-media-query
+import { useMediaQuery } from '../useMediaQuery'
+
+import { useUpdateEffect } from '..'
+
+const COLOR_SCHEME_QUERY = '(prefers-color-scheme: dark)'
+
+// TODO: use enum
+type TernaryDarkMode = 'system' | 'dark' | 'light'
+interface UseTernaryDarkModeOutput {
+  isPaletteDarkMode: boolean
+  ternaryModeCode: TernaryDarkMode // change icon
+  toggleTernaryDarkMode: () => void
+}
+
+function useTernaryDarkMode(): UseTernaryDarkModeOutput {
+  /* ======== HELPERS ======== */
+  function useDarkModeOS(): boolean {
+    return useMediaQuery(COLOR_SCHEME_QUERY)
+    // Prevents SSR issues, old test failed const mQuery = useMediaQuery(COLOR_SCHEME_QUERY);
+    // if (typeof window !== 'undefined') {
+    //   if (window.hasOwnProperty('matchMedia')) {
+    //     return window.matchMedia(COLOR_SCHEME_QUERY).matches
+    //   }
+    // }
+    // return !!defaultValue
+  }
+  /* ======== REACT ======== */
+  const [isDarkOS, setDarkOS] = useState<boolean>(useDarkModeOS())
+  const [ternaryMode, setTernaryMode] = useLocalStorage<TernaryDarkMode>(
+    'ternaryDarkMode',
+    'system',
+  )
+  const [isDarkMode, setDarkMode] = useState<boolean>(true)
+
+  // Update darkMode if os prefers changes
+  useUpdateEffect(() => {
+    setDarkOS(isDarkOS)
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [isDarkOS])
+  // useEffect(() => {
+  //   const handler = () => setDarkOS(getPrefersDarkMode())
+  //   const matchMedia = window.matchMedia(COLOR_SCHEME_QUERY)
+  //   matchMedia.addEventListener('change', handler)
+  //   return () => {
+  //     matchMedia.removeEventListener('change', handler)
+  //   }
+  //   // eslint-disable-next-line react-hooks/exhaustive-deps
+  // }, [])
+
+  useEffect(() => {
+    // for setIsPaletteDarkMode
+    if (ternaryMode === 'dark') {
+      setDarkMode(true)
+    } else if (ternaryMode === 'system') {
+      setDarkMode(isDarkOS)
+    } else if (ternaryMode === 'light') {
+      setDarkMode(false)
+    } else {
+      console.error('ERR reading palette mode')
+    }
+  }, [ternaryMode, isDarkOS])
+
+  return {
+    isPaletteDarkMode: isDarkMode,
+    ternaryModeCode: ternaryMode,
+    toggleTernaryDarkMode: () => {
+      setTernaryMode((currMode: TernaryDarkMode) =>
+        currMode === 'dark'
+          ? 'light'
+          : currMode === 'light'
+          ? 'system'
+          : 'dark',
+      )
+    },
```

Comment:
About the SSR problem, I'm going to just apply `useMediaQuery`, since the `useDarkMode` is using it as well. We will only need to change one place to fix all.

---

The design was for using one button for the minimalist style. You can check it on the top bar of https://www.snippethub.dev/ to see how it works. For me this is better than using an extra button. It keeps the dark/light toggle language and adds a new function. This is how I thought when I wrote the function.

If we need another toggle, its better to use the "use dark mode" and a "use system color override".